app/lib/frontend/backend.dart

```diff
@@ -159,12 +160,66 @@ class Backend {
     assert(repository.supportsDownloadUrl);
     return repository.downloadUrl(package, version);
   }
+
+  /// Stores a verification entry in the Datastore and returns its id that can
+  /// be used in client-communication, e.g. sending via e-mail.
+  ///
+  /// Duplicate detection relies on the serialized JSON form of [parameters],
+  /// and it is expected that callers will order the keys consistently.
+  /// When a duplicate is detected, the return value becomes `null`.
+  Future<String> createVerification(
+      String action, Map<String, dynamic> parameters,
+      {Duration expires = const Duration(days: 1)}) async {
+    final id = _uuid.v4().toString();
+    final now = new DateTime.now().toUtc();
+    final verification = new models.Verification()
+      ..parentKey = db.emptyKey
+      ..id = id
+      ..created = now
+      ..expires = now.add(expires)
+      ..action = action
+      ..parameters = parameters
+      ..updateHash();
+
+    final query = db.query<models.Verification>()
+      ..filter('dedupHash =', verification.dedupHash);
+    await for (var v in query.run()) {
+      if (!v.isActive()) continue;
+      // TODO: match parameters too for 100% certainty
+      // The chance of mis-classification is low, but we should eventually address it.
+      if (v.action == action) {
+        return null;
+      }
+    }
+
+    await db.commit(inserts: [verification]);
+    return id;
+  }
+
+  /// Updates the verification entry (identified by [action] and [id]), and
+  /// return it parameters (or null if it has been expired).
+  Future<Map<String, dynamic>> confirmVerification(
```

Comment:
You should document that any action taken following this verification should be idempotent.

We should probably allow this to go through, if the user tries to verify within 5 minutes.

Reply:
Is this in case the verification process is aborted for some reason? Yeah, an additional 5-10 minutes window would be okay I guess.